server/src/main/java/org/opensearch/action/search/TransportSearchAction.java

```diff
@@ -442,6 +444,11 @@ private void executeRequest(
             );
             searchRequestContext.getSearchRequestOperationsListener().onRequestStart(searchRequestContext);
 
+            // At this point either the QUERY_GROUP_ID header will be present in ThreadContext either via ActionFilter
+            // or HTTP header (HTTP header will be deprecated once ActionFilter is implemented)
+
+            ((QueryGroupTask) task).setQueryGroupId(threadPool.getThreadContext());
```

Comment:
nit: Can you add `instanceof` check to avoid any typecast exceptions? We can log error if not as expected